File: C++/代码随想录/Linked_List/707. Design Linked List.cpp
// 707. Design Linked List
// https://leetcode.com/problems/design-linked-list/

#include "LinkedListUtils.hpp"

class MyLinkedList {
private:
    ListNode* head;
public:
    MyLinkedList() {
        head = nullptr;
    }

    int get(int index) {
        ListNode* current = head;
        int i = 0;
        while (current != nullptr && i < index) {
            current = current->next;
            i++;
        }
        if (current == nullptr) {
            return -1;
        }
        return current->val;
    }
    
    void addAtHead(int val) {
        ListNode* newNode = new ListNode(val);
        newNode->next = head;
        head = newNode;
    }

    void addAtTail(int val) {
        ListNode* newNode = new ListNode(val);
        if (head == nullptr) {
            head = newNode;
        } else {
            ListNode* current = head;
            while (current->next != nullptr) {
                current = current->next;
            }
            current->next = newNode;
        }
    }
    
    void addAtIndex(int index, int val) {
        if (index < 0) {
            return;
        }
        if (index == 0) {
            addAtHead(val);
            return;
        }
        ListNode* newNode = new ListNode(val);
        ListNode* current = head;
        int i = 0;
        while (current != nullptr && i < index - 1) {
            current = current->next;
            i++;
        }
        if (current == nullptr) {
            return;
        }
        newNode->next = current->next;
        current->next = newNode;
    }
    
    void deleteAtIndex(int index) {
        if (index < 0) {
            return;
        }
        if (index == 0) {
            if (head != nullptr) {
                ListNode* temp = head;
                head = head->next;
                delete temp;
            }
            return;
        }
        ListNode* current = head;
        int i = 0;
        while (current != nullptr && i < index - 1) {
            current = current->next;
            i++;
        }
        if (current == nullptr || current->next == nullptr) {
            return;
        }
        ListNode* temp = current->next;
        current->next = current->next->next;
        delete temp;
    }
};

int main() {
    MyLinkedList linkedList;
    linkedList.addAtHead(1);
    linkedList.addAtTail(3);
    linkedList.addAtIndex(1, 2);
    std::cout << linkedList.get(1) << std::endl;
    linkedList.deleteAtIndex(1);
    std::cout << linkedList.get(1) << std::endl;

    return 0;
}

File: C++/代码随想录/Linked_List/707. Design Linked List.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class MyLinkedList:
    def __init__(self):
        self.size = 0
        self.head = ListNode(0)  # 虚拟头节点

    def get(self, index: int) -> int:
        if index < 0 or index >= self.size:
            return -1
        current = self.head
        for _ in range(index + 1):
            current = current.next
        return current.val

    def addAtHead(self, val: int) -> None:
        self.addAtIndex(0, val)

    def addAtTail(self, val: int) -> None:
        self.addAtIndex(self.size, val)

    def addAtIndex(self, index: int, val: int) -> None:
        if index > self.size:
            return
        if index < 0:
            index = 0
        self.size += 1
        pred = self.head
        for _ in range(index):
            pred = pred.next
        to_add = ListNode(val)
        to_add.next = pred.next
        pred.next = to_add

    def deleteAtIndex(self, index: int) -> None:
        if index < 0 or index >= self.size:
            return
        self.size -= 1
        pred = self.head
        for _ in range(index):
            pred = pred.next
        pred.next = pred.next.next

# 测试用例
linked_list = MyLinkedList()
linked_list.addAtHead(1)
linked_list.addAtTail(3)
linked_list.addAtIndex(1, 2)  # 链表变为1->2->3
print(linked_list.get(1))     # 返回2
linked_list.deleteAtIndex(1)  # 现在链表是1->3
print(linked_list.get(1))     # 返回3
